### PR TITLE
feat: add should_use_tensor_core

### PR DESCRIPTION
**Pull request by zhyncs**
## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by zhyncs** [comment]
follow-up https://github.com/sgl-project/sglang/pull/2169

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 1]
```diff
@@ -45,19 +44,14 @@ class FlashInferAttnBackend(AttentionBackend):
...
-            )
-        else:
-            if not _grouped_size_compiled_for_decode_kernels(
-                model_runner.model_config.num_attention_heads // model_runner.tp_size,
-                model_runner.model_config.get_num_kv_heads(model_runner.tp_size),
-            ):
-                self.decode_use_tensor_cores = True
-            else:
-                self.decode_use_tensor_cores = False
+        env_override = os.environ.get("SGLANG_FLASHINFER_USE_TENSOR_CORE")
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py]: also move this env check `should_use_tensor_cores`

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/utils.py [hunk 2]
```diff
@@ -987,3 +987,37 @@ def direct_register_custom_op(
...
+        return env_override.lower() == "true"
+
+    # Calculate GQA group size
+    gqa_group_size = num_attention_heads // num_kv_heads
+
+    # Determine based on dtype and GQA group size
+    if kv_cache_dtype == torch.float8_e4m3fn:
+        return True
+    elif kv_cache_dtype in (torch.float16, torch.half, torch.bfloat16):
+        return gqa_group_size > 4
```
Comment (merrymercy) [inline python/sglang/srt/utils.py:1156]: I think this is reverted due to a regression bug https://github.com/sgl-project/sglang/pull/1511/files.

Has it been fixed in v0.1.6?

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/utils.py (hunk 2 above)
Comment (zhyncs) [reply python/sglang/srt/utils.py:1156]: This should be fixed in v0.2

File: python/sglang/srt/layers/attention/flashinfer_backend.py (hunk 1 above)
Comment (zhyncs) [reply python/sglang/srt/layers/attention/flashinfer_backend.py]: done

File: python/sglang/srt/utils.py (hunk 2 above)
Comment (zhyncs) [reply python/sglang/srt/utils.py:1156]: flashinfer v0.1.6 nightly should also work

**Review by merrymercy — APPROVED** [review]
Let us get this in first!

File: python/sglang/srt/utils.py [hunk 3]
```diff
@@ -1108,3 +1108,50 @@ def cuda_device_count_stateless() -> int:
...
+        bool: Whether to use tensor cores
+    """
+    # Try to use environment variable and dtype-based logic first
+    env_override = os.environ.get("SGLANG_FLASHINFER_USE_TENSOR_CORE")
+    if env_override is not None:
+        return env_override.lower() == "true"
+
+    # Try to use _grouped_size_compiled_for_decode_kernels if available
+    try:
+        from flashinfer.decode import _grouped_size_compiled_for_decode_kernels
```
Comment (merrymercy) [inline python/sglang/srt/utils.py:1137]: comment that this is for flashinfer <=0.1.6. Otherwise, there is an accuracy bug

Comment (zhyncs) [reply python/sglang/srt/utils.py:1137]: ok

Comment (zhyncs) [reply python/sglang/srt/utils.py:1137]: done

**Pull request merged**